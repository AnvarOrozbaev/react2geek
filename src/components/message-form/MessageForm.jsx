import { useState, useRef } from 'react';
import './MessageForm.scss';

export const MessageForm = ({ onSendMessage }) => {
  const [value, setValue] = useState('');
  const inputRef = useRef();

  const handleChange = (e) => {
    setValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (value.length > 0) {
      onSendMessage({
        author: 'me',
        text: value,
      });

      setValue('');
    }
  };

  return (
    <form className="  message-form" onSubmit={handleSubmit}>
      <input
        className="message-form__input"
        ref={inputRef}
        type="text"
        value={value}
        placeholder="сообщение"
        onChange={handleChange}
      />
      <input className="message-form__btn" type="submit" />
    </form>
  );
};
