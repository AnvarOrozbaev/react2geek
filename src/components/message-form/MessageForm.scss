$primary:   #864f12;

.message-form{
  margin-top: 2rem;
  display: flex;
  align-items: center;
  
&__input{
  border-radius: 10px;
  border: 1px solid $primary;
  border-right: none;
  border-top-right-radius: 0 ;
  border-bottom-right-radius: 0 ;
  outline: none;
  height: 1.2em;
}
&__btn{
  border-radius: 10px;
  border: 1px solid $primary;
  border-top-left-radius: 0 ;
  border-bottom-left-radius: 0 ;
  &:hover{
    background-color: #ffffff;
    color: $primary;
  }
  
}
}
.--primary{
  background-color: $primary;
  color:  #ffffff;
}
.--secondary{
  background-color:  #ffffff;
  color: $primary;
}