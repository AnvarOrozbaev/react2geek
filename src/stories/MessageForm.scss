$primary:   #864f12;

.message-form{
  margin-top: 2rem;
  
&__input{
  border-radius: 10px;
  border: 2px solid $primary;
  border-right: none;
  border-top-right-radius: 0 ;
  border-bottom-right-radius: 0 ;
}
&__btn{
  border-radius: 10px;
  border: 2px solid $primary;

  border-top-left-radius: 0 ;
  border-bottom-left-radius: 0 ;
  &:hover{
    background-color: #ffffff;
    color: $primary;
  }
  
}
}
.--primary{
  background-color: $primary;
  color:  #ffffff;
}
.--secondary{
  background-color:  #ffffff;
  color: $primary;
}