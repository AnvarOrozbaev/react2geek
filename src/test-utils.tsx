import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import chatsReducer from './pages/chats/chatsSlice';
import profileReducer from './pages/profile/profileSlice';
import articlesReducer from './pages/articles/articlesSlice';
type Props = {
  children?: JSX.Element;
};
function render(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        chatsState: chatsReducer,
        profile: profileReducer,
        articles: articlesReducer,
      },
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }: Props) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { render };
