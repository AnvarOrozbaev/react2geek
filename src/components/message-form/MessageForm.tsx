import React, { useEffect } from 'react';
import { useState, useRef, memo } from 'react';
import InputBase from '@mui/material/InputBase';
import './MessageForm.scss';
import { useSelector } from 'react-redux';
import { selectSelectedId } from '../../pages/chats/chatsSlice'
interface MessageFormProps {
  primary?: boolean;
  onSendMessage: (text: string) => void;
}
export const MessageForm = memo<MessageFormProps>(
  ({ primary, onSendMessage }) => {
    const mode = primary ? '--secondary' : '--primary';
    const [value, setValue] = useState<string>('');
    const selectedId = useSelector(selectSelectedId)
    const inputRef = useRef<HTMLInputElement>(null);
    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      setValue(e.target.value);
    };
    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      if (value.length === 0) {
        return;
      }
      onSendMessage(value);
      setValue('');
      if (inputRef.current) {
        inputRef.current.focus();
      }
    };
    useEffect(()=>{
      if (inputRef.current) {
        inputRef.current.focus();
      }
    },[selectedId])

    return (
      <form
        className="message-form"
        onSubmit={handleSubmit}
      >
        <InputBase
          className="message-form__input"
          inputRef={inputRef}
          autoFocus={true}
          type="text"
          value={value}
          placeholder="сообщение"
          onChange={handleChange}
        />
        <input
          className={['message-form__btn', mode].join(' ')}
          type="submit"
          data-testid="message-form__btn"
        />
      </form>
    );
  }
);
