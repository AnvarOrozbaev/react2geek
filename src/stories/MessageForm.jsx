import './MessageForm.scss';

export const MessageForm = ({ primary }) => {
  const mode = primary ? '--primary' : '--secondary';
  return (
    <form className="message-form">
      <input
        className="message-form__input"
        type="text"
        placeholder="сообщение"
      />
      <input className={['message-form__btn', mode].join('')} type="submit" />
    </form>
  );
};
