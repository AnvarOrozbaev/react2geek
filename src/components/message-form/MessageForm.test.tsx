import React from 'react';
import { MessageForm } from './MessageForm';
import { render, fireEvent, screen } from '../../test-utils';
import '@testing-library/jest-dom';

describe('MessageForm', () => {
  it('render component', () => {
    const mockHandler = jest.fn();
    render(<MessageForm onSendMessage={mockHandler} />);
  });
  it('render with snapshot', () => {
    const mockHandler = jest.fn();
    const { asFragment } = render(<MessageForm onSendMessage={mockHandler} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('render btn', async () => {
    const mockHandler = jest.fn();
    render(<MessageForm onSendMessage={mockHandler} />);
    const btn = screen.getByTestId('message-form__btn');
    expect(btn).toBeInTheDocument();
  });
  it('render input', async () => {
    const mockHandler = jest.fn();
    const { getByPlaceholderText } = render(
      <MessageForm onSendMessage={mockHandler} />
    );
    const input = getByPlaceholderText(/сообщение/i);
    expect(input).toBeInTheDocument();
  });
  it('form submit', async () => {
    const mockHandler = jest.fn();
    const { getByPlaceholderText } = render(
      <MessageForm onSendMessage={mockHandler} />
    );
    const btn = screen.getByTestId('message-form__btn');
    const input = getByPlaceholderText(/сообщение/i);

    fireEvent.change(input, { target: { value: 'hello' } });
    fireEvent.click(btn);
    expect(mockHandler).toBeCalled();
  });
  it('form behavior when input is empty', async () => {
    const mockHandler = jest.fn();
    const { getByPlaceholderText } = render(
      <MessageForm onSendMessage={mockHandler} />
    );
    const btn = screen.getByTestId('message-form__btn');
    const input = getByPlaceholderText(/сообщение/i);

    fireEvent.change(input, { target: { value: '' } });
    fireEvent.click(btn);
    expect(mockHandler).toHaveBeenCalledTimes(0);
  });
});
