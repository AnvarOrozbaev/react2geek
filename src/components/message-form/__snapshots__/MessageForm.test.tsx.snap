// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm render with snapshot 1`] = `
<DocumentFragment>
  <form
    class="message-form"
  >
    <div
      class="MuiInputBase-root MuiInputBase-colorPrimary Mui-focused message-form__input css-1le28he-MuiInputBase-root"
    >
      <input
        class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
        placeholder="сообщение"
        type="text"
        value=""
      />
    </div>
    <input
      class="message-form__btn --primary"
      data-testid="message-form__btn"
      type="submit"
    />
  </form>
</DocumentFragment>
`;
